### DISCUSSION TITLE
What do branch colours mean?

### DISCUSSION THREAD

**Discussion by unikitty37** [jesseduffield/lazygit#3680] [Q&A] [answered]
Following on from #1622, what do branch colours mean?

![image](https://github.com/jesseduffield/lazygit/assets/139183/c591ba33-c765-4f8f-a8ba-4e1d7c6197a6)

The branch starting `feature/` is green, and I don't know why.

Similarly, `master` is considered a main branch according to the default settings for `git.mainBranches`, but it doesn't show any differently. I'm not sure if it _should_ — it seems like that setting is intended to highlight commits that are part of a main branch — though highlighting main branches might be useful. Of course, there's a way to do that…

If I set `gui.branchColors` to give the branches a different colour to those branches, `master` shows in the custom colour, but the mystery greenness overrides my setting for the feature branch and my custom colour is ignored.

What _does_ a green branch name mean? Is it just that the branch was pulled recently? If so, is there a way to disable that? (I can already see it's recent from the "23h", and having override a custom colour really doesn't help, as I configured the custom colours for a reason :)

**Comment by stefanhaller** [comment, marked as answer]
There's some [very old code](https://github.com/jesseduffield/lazygit/blob/3af545daf7cf6458e8efd324012047ce688f08e6/pkg/gui/presentation/branches.go#L127) that has a hard-coded mapping of branch prefixes to colors.

I find this pretty annoying and would vote for removing this code. If people do want to have a color scheme like this, they should configure their `gui.branchColors` accordingly, but for those who don't, the only way to turn it off right now is to say
```
gui:
  branchColors:
    feature: default
    bugfix: default
    hotfix: default
```
(I don't use these branch prefixes myself, but I do occasionally check out other people's branches who do, and I don't want these to stand out in my branches list.)

I'm surprised, however, that you say "the mystery greenness overrides my setting for the feature branch and my custom colour is ignored"; it should be the other way round, if you have a custom color it should win.

As for coloring the main branches specially, I like the idea. I wonder if it should be a separate setting though (e.g. `gui.mainBranchesColor: yellow`). It's not totally clear which setting should win if a main branch matches one of the custom branch colors; also, it's not totally trivial to implement, because you can't just match the local branch name against the main branches list. For example, I have repositories where the main branch is `origin/master`, but the corresponding local branch which should be colored is `master-upstream`, because I also have `master-myfork`. It might be easier to let users color these via `gui.branchColors` though.

**Comment by unikitty37** [comment]
Thanks — interestingly, if I have this, `feature/foo` is still white, not red. Swapping the order of the `feature` and `feature/foo` lines doesn't change that.

```yaml
gui:
  branchColors:
    feature: default
    "feature/foo": red
    bugfix: default
    hotfix: default
```

**Reply by stefanhaller** [reply]
Yeah, the way the branchColors feature is implemented is different from what you would intuitively expect. You'd probably expect to loop over the entries in the array, seeing which one is a prefix of the branch name, and then either take the first one found, or the longest one. But that's not how it works; it splits the branch name at the first `/`, and then looks for an exact match of what's to the left of it. This means that `feature/foo` can't possibly be found, it will always only find `feature`.

**Reply by unikitty37** [reply]
Rats. I guess wildcard support would be nice (or regular expressions 😈) but that's definitely a breaking change.
